constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/IntervalMap.java

```diff
@@ -0,0 +1,500 @@
+import static gov.nasa.jpl.aerie.constraints.time.Interval.interval;
+
+public final class IntervalMap<V> implements Iterable<Segment<V>> {
+
+  private final IntervalAlgebra alg;
+
+
+  //[0, 3) [3] (3,5) -> would be 3 windows, one is zero width, we ARE allowing this.
+
+  // INVARIANT: `intervals` is list of non-empty, non-overlapping intervals in ascending order.
+  // INVARIANT: If two adjacent intervals abut exactly (e.g. [0, 3), [3, 5]), their values are non-equal.
+  private final List<Segment<V>> segments;
+
+  public IntervalMap() {
+    this.segments = new ArrayList<>();
+    this.alg = new IntervalAlgebra(Interval.FOREVER);
+  }
+
+  public IntervalMap(final IntervalAlgebra algebra, final List<Segment<V>> segments) {
+    this(algebra);
+    for (final var segment: segments) {
+      this.setInternal(segment.interval(), segment.value(), 0);
+    }
+  }
+
+  public IntervalMap(final List<Segment<V>> segments) {
+    this.alg = new IntervalAlgebra();
+    this.segments = new ArrayList<>(segments.size());
+    for (final var segment: segments) {
+      this.setInternal(segment.interval(), segment.value(), 0);
+    }
+  }
+
+  public IntervalMap(final IntervalAlgebra algebra) {
+    this.alg = algebra;
+    this.segments = new ArrayList<>();
+  }
+
+  public IntervalMap(final Interval interval, final V value) {
+    this.alg = new IntervalAlgebra(Interval.FOREVER);
+    this.segments = List.of(Segment.of(interval, value));
+  }
+
+  @SafeVarargs
+  public IntervalMap(final Segment<V>... segments) {
+    this();
+    for (final var segment: segments) {
+      this.setInternal(segment.interval(), segment.value(), 0);
+    }
+  }
+
+  public IntervalMap(final IntervalAlgebra algebra, final Interval key, final V value) {
+    this(algebra, new ArrayList<>(List.of(Segment.of(Objects.requireNonNull(key), Objects.requireNonNull(value)))));
+  }
+
+  public IntervalMap(final List<Interval> intervals, final V value) {
+    this(
+        new IntervalAlgebra(Interval.FOREVER),
+        intervals.stream().map($ -> Segment.of($, value)).toList()
+    );
+  }
+
+  // Copy constructor
+  public IntervalMap(final IntervalMap<V> other) {
+    this(other.alg, new ArrayList<>(other.segments));
+  }
+
+  public IntervalMap(final V value) {
+   this(Interval.FOREVER, value);
+  }
+
+  public IntervalMap<V> set(final Interval interval, final V value) {
+    return set(new IntervalMap<>(interval, value));
+  }
+
+  public IntervalMap<V> set(final List<Interval> intervals, final V value) {
+    return set(new IntervalMap<>(intervals, value));
+  }
+
+  public IntervalMap<V> set(final IntervalMap<V> other) {
+    final var result = new IntervalMap<>(this);
+    int index = 0;
+    for (final var segment: other) {
+      index = result.setInternal(segment.interval(), segment.value(), index);
+    }
+    return result;
+  }
+
+  int setInternal(Interval interval, final V value, int index) {
```

Comment:
🔶 What are the invariants on this `int index`? What does it mean on input, and what does it mean on output? I'm a little lost, and would appreciate a comment here.

EDIT: upon reading to the bottom of the function, I found this comment:
```java
// now, everything left of `index` is strictly left of `interval`,
// and everything right of `index` is strictly right of `interval`,
```
which helps a little bit, but still doesn't help me understand what the caller is supposed to do about it.

Reply:
I'll document it a bit. To my understanding (Pranav wrote it so I'm not completely sure), the caller guarantees that all segments before `index` are strictly before `interval` and do not need to be checked for overwriting and coalescing. The return value is a new index, and `setInternal` guarantees that it will be a valid argument if you call `setInternal` again with another interval that is strictly after the first. This is useful when assigning several sorted non-overlapping segments at once.

```java
public IntervalMap<V> set(final IntervalMap<V> other) {
  final var result = new IntervalMap<>(this);
  int index = 0;
  for (final var segment: other) {
    index = result.setInternal(segment.interval(), segment.value(), index);
  }
  return result;
}
```

Ultimately its just an internal optimization, and supplying `index = 0` to `setInternal` will always be valid.